ferry.go

```diff
@@ -1021,3 +1030,88 @@ func (f *Ferry) checkConnectionForBinlogFormat(db *sql.DB) error {
 
 	return nil
 }
+
+func (f *Ferry) UpdateBatchSizes() error {
+	schemaTablesMap := map[string][]string{}
+
+	for _, table := range f.Tables {
+		if _, found := f.Config.DataIterationBatchSizePerTableOverride.TableOverride[table.Schema][table.Name]; found {
+			continue
+		}
+		schemaTablesMap[table.Schema] = append(schemaTablesMap[table.Schema], table.Name)
+	}
+
+	for schemaName, tableNames := range schemaTablesMap {
+		if _, found := f.Config.DataIterationBatchSizePerTableOverride.TableOverride[schemaName]; !found {
+			f.Config.DataIterationBatchSizePerTableOverride.TableOverride[schemaName] = map[string]uint64{}
+		}
+		query := fmt.Sprintf(
+			`SELECT T1.TABLE_NAME, AVG_ROW_LENGTH, MAX_BATCH_SIZE
+					FROM information_schema.TABLES T1
+						JOIN (SELECT TABLE_NAME, TABLE_SCHEMA, floor(%d / count(*)) MAX_BATCH_SIZE
+						FROM information_schema.COLUMNS WHERE TABLE_SCHEMA = '%s' AND TABLE_NAME in (%s)
+						GROUP BY table_name) T2 USING (TABLE_SCHEMA, TABLE_NAME);`,
+			MySQLNumParamsLimit,
+			schemaName,
+			"'"+strings.Join(tableNames, "', '")+"'",
+		)
+		rows, err := f.SourceDB.Query(query)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var avgRowLength, maxBatchSize int
+			var tableName string
+			err = rows.Scan(&tableName, &avgRowLength, &maxBatchSize)
+			if err != nil {
+				return err
+			}
+			// MySQL has an upper limit of 2^16 for num_params for prepared statements. This will cause an error when
+			// we attempt to execute a INSERT statement with > 65535 parameters.
+			// If batchSize > 65535/num_columns, we will use 65535/num_columns
+			batchSize := f.calculateBatchSize(avgRowLength)
+			if batchSize >= maxBatchSize {
+				batchSize = maxBatchSize
+			}
+			f.Config.DataIterationBatchSizePerTableOverride.TableOverride[schemaName][tableName] = uint64(batchSize)
+		}
+
+	}
+	return nil
+}
+
+func (f *Ferry) calculateBatchSize(rowSize int) int {
+	batchSizePoints := f.Config.DataIterationBatchSizePerTableOverride
```

Comment:
This method really feels like it should be attached to the DataIterationBatchSizePerTableOverride struct as opposed to be on the ferry. In fact, the first line of this function assigns `DataIterationBatchSizePerTableOverride` as a variable and no where else do we need `f` anymore.

Reply:
Moved into `DataIterationBatchSizePerTableOverride` struct